### ISSUE TITLE
User Statistics: passing additional state into closure

### ISSUE THREAD

**Issue by tabokie** [google/benchmark#848]
Currently I am trying to create a complex user statistics by `ComputeStatistics`, and the computing logic is related to state parameter (coroutine number), passing a `State` reference to closure will nicely solve this problem.

Is there any other ways to achieve this purpose? If not is it okay to change the `ComputeStatistics` (in which case i will file a pr soon)?

**Comment by dominichamon** [comment]
Can you add a bit more detail about what you're trying to do?

Why do you think the coroutine number should be part of the statistics you're reporting?

**Comment by tabokie** [comment]
> Can you add a bit more detail about what you're trying to do?
> 
> Why do you think the coroutine number should be part of the statistics you're reporting?

e.g. computing the average coroutine time, which equals `iteration time / total coroutine`

**Comment by LebedevRI** [comment]
Can you show a snippet?
What's `total coroutine` ?

**Comment by tabokie** [comment]
> Can you show a snippet?
> What's `total coroutine` ?

like this one, though a bit trivial
```c++
BENCHMARK_DEFINE_F(CoroTest, Basic)(benchmark::State& state) {
  CoSched pool;
  uint32_t num_coroutine = state.range(0);
  uint32_t num_ctx_switch = 0;
  for (auto _ : state) {
    pool.submit(num_coroutine, [] {
      for (int i = 0; i < 100; i++) {
        this_routine::yield();
      }
    });
    pool.wait();
    num_ctx_switch += pool.num_context_switch();
  }
  state.counters["ctx_switch"] = num_ctx_switch;
}

BENCHMARK_REGISTER_F(CoroTest, Basic)->Ranges(1, 100)
->ComputeStatistics("avg_ctx_switch", [](const benchmark::State& state, std::vector<double>& v) {
  return std::accumulate(v.begin(), v.end(), 0.0) / state.counters["ctx_switch"];
});
```

**Comment by LebedevRI** [comment]
So it's just total time divided by final `state.counters["ctx_switch"]` ?
This is not motivational-enough to change `ComputeStatistics()`.

You want a simple custom counter instead:
The missing piece here is https://github.com/google/benchmark/issues/781 (sigh, i should just implement it already :/),
then it should just be
```c++
// avg_ctx_switch = 1 / (num_ctx_switch / time)
state.counters["avg_ctx_switch"] = benchmark::Counter(num_ctx_switch,
                                                      benchmark::Counter::kIsRate |
                                                      benchmark::Counter::KInvert);
```

**Comment by tabokie** [comment]
That's nice! never thought of using counter before.
 Will you pr soon? Or i could do it later.

**Comment by LebedevRI** [comment]
#850, cheers.

**Comment by LebedevRI** [comment]
(opening so this gets properly closed via commit)

**Comment by LebedevRI** [comment]
Fixed by #850. (github didn't parse multiple issues at once..)
